src/main/java/org/apache/commons/io/input/TrailerInputStream.java

```diff
@@ -0,0 +1,180 @@
+            readIntoBuffer = len;
+            // b: [1234], trailer: [----56789] --> b: [1234], trailer: [56789----]
+            System.arraycopy(this.trailer, len, this.trailer, 0, this.trailer.length - len);
+            // b: [1234], trailer: [56789----] --> b: [1234], trailer: [56789abcd]
+            read = IOUtils.read(this.source, this.trailer, this.trailer.length - len, len);
+        } else {
+            // 1 or 2 IOUtils.read calls to source, copies this.trailer.length bytes
+            // trailer smaller than b, so need to read into b and trailer
+            // b: [---------], t: [1234] --> b: [1234-----], t: [----]
+            System.arraycopy(this.trailer, 0, b, off, this.trailer.length);
+            // b: [1234-----], t: [----] --> b: [1234abcde], t: [----]
+            read = IOUtils.read(
+                            this.source, b, off + this.trailer.length, len - this.trailer.length);
+            readIntoBuffer = this.trailer.length + read;
+            // b: [1234abcde], t: [----] --> b: [1234abcde], t: [fghi]
+            if (read == len - this.trailer.length) { // don't try reading data when stream source EOF
+                read += IOUtils.read(this.source, this.trailer);
+            }
+        }
+        // if less data than requested has been read, the trailer buffer is not full
+        // --> need to fill the trailer with the last bytes from b
+        // (only possible if we reached EOF)
+        // second step: ensure that trailer is completely filled with data
+        // overview - b: [abcdefghi], t: [jk--] --> b: [abcdefg--], t: [hijk]
+        final int underflow = Math.min(len - read, this.trailer.length);
+        if (underflow > 0) {
+            // at most this.trailer.length are copied to fill the trailer buffer
+            if (underflow < this.trailer.length) {
+                // trailer not completely empty, so move data to the end
+                // b: [abcdefghi], t: [jk--] --> b: [abcdefghi], t: [--jk]
+                System.arraycopy(
+                        this.trailer, 0, this.trailer, underflow, this.trailer.length - underflow);
+            }
+            // fill trailer with last bytes from b
+            // b: [abcdefghi], t: [--jk] --> b: [abcdefg--], t: [hijk]
+            System.arraycopy(b, off + readIntoBuffer - underflow, this.trailer, 0, underflow);
+        }
+        // IOUtils.read reads as many bytes as possible, so reading 0 bytes means EOF.
+        // Then, we have to mark this.
+        return read == 0 && len != 0 ? IOUtils.EOF : read;
+    }
+
+    @Override
+    public int available() throws IOException {
+        return this.source.available();
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            this.source.close();
+        } finally {
+            super.close();
+        }
+    }
+
+    public int getTrailerLength() {
+        return this.trailer.length;
+    }
+
+    public byte[] copyTrailer() {
```

Comment:
I think one method to getTrailer() should suffice and would be noticeably simpler. I don't expect trailers to be all that long.

Reply:
Done.